File: python3/path_sum.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""
class Solution:
    def hasPathSum(self, root: TreeNode, s: int) -> bool:
        if not root:
            return False

        stack = [(root, 0)]

        while stack:
            node, count = stack.pop()
            if not node.left and not node.right:
                if count + node.val == s:
                    return True
            if node.left:
                stack.append((node.left, count + node.val))
            if node.right:
                stack.append((node.right, count + node.val))

        return False

File: golang/path_sum.go
package main

import "golang/model"

type nodeStack struct {
	Node *model.TreeNode
	Path int
}

func hasPathSum(root *model.TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	stack := []nodeStack{
		nodeStack{
			Node: root,
			Path: 0,
		},
	}

	for len(stack) > 0 {
		var temp []nodeStack
		for _, ns := range stack {
			node := ns.Node
			path := ns.Path

			if node.Left == nil && node.Right == nil {
				if path+node.Val == targetSum {
					return true
				}
			}
			if node.Left != nil {
				temp = append(temp, nodeStack{Node: node.Left, Path: path + node.Val})
			}
			if node.Right != nil {
				temp = append(temp, nodeStack{Node: node.Right, Path: path + node.Val})
			}
		}
		stack = temp
	}

	return false
}
